### PR TITLE
add returnType to UDFRegistration methods

### PR DESCRIPTION
**Pull request by Velociraptor115**
This PR adds new methods to support schema for Row return type in UDF registration. This implements the feature request of #455 

The current UDF registration API does not accept a schema parameter. This prevents us from registering an UDF like
```
var schema = new StructType(new[]
{
    new StructField("col1", new IntegerType())
});
Func<Column, Column> udf = Udf<string>(
    str => new GenericRow(new object[] { 111 }), schema);

// Though I can get `udf` as a Func<Column, Column>, I can't register it,
// since the spark.Udf().Register* functions do not accept the `returnType` param
```
even though the underlying API supports it.

This PR adds new methods for UDF registration with the returnType param for functions that return Row.
This allows us to register such UDFs like
```
var schema = new StructType(new[]
{
    new StructField("col1", new IntegerType())
});
_spark.Udf().Register<string>(
    "my_udf",
    str => new GenericRow(new object[] { 111 }),
    schema);
```

### CODE REVIEW

**Comment by elvaliuliuliu** [comment]
@Velociraptor115 : Thanks for your PR and contribution! Can you please fix the failure and also add a test for this feature?

**Comment by Velociraptor115** [comment]
> 
> 
> @Velociraptor115 : Thanks for your PR and contribution! Can you please fix the failure and also add a test for this feature?

Thanks for your quick response.
Would be glad to add tests.

Can you point me to the best place to add them.
I did a quick check for the existing UDF registration methods and couldn't find any places where they were referred, except a couple of places in the examples

**Comment by imback82** [comment]
Thanks @Velociraptor115 for your first PR! Can you update the description with an example that is not working that this PR proposes to fix. And you can convert that example to tests.

**Comment by Velociraptor115** [comment]
I have added tests in `UdfComplexTypesTests.cs`

I did not have a specific example in mind to use for the tests.
My current use case involves registering an UDF in .NET (as I need to use a .NET library)
and invoking that from Scala (since the primary codebase is in Scala)

I have however reused the same tests from `TestUdfWithReturnAsRowType`, rewriting the UDFs to be registered and called from SQL instead.

File: src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs [hunk 1]
```diff
@@ -320,5 +320,165 @@ public void TestUdfWithReturnAsRowType()
...
+                    "udf5",
+                    row => row.GetAs<string>(1));
+
+                Row[] rows =
+                    _spark.Sql("SELECT udf5(udf4(name)) FROM people")
+                    .Collect()
+                    .ToArray();
+                Assert.Equal(3, rows.Length);
+
+                var expected = new[] { "Michael", "Andy", "Justin" };
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs]: nit: Probably existing code, but it should be `var expected = new string[] { ... }`;

File: src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs [hunk 2]
```diff
@@ -320,5 +320,165 @@ public void TestUdfWithReturnAsRowType()
...
+
+                _spark.Udf().Register<string>(
+                    "udf1",
+                    str => new GenericRow(new object[] { 1, "abc" }),
+                    schema);
+
+                Row[] rows =
+                    _spark.Sql("SELECT udf1(name) AS col FROM people")
+                    .Collect()
+                    .ToArray();
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs]: Either
```C#
Row[] rows =
    _spark.Sql("SELECT udf1(name) AS col FROM people")
        .Collect()
        .ToArray();
```
or
```C#
Row[] rows =_spark.Sql("SELECT udf1(name) AS col FROM people")
    .Collect()
    .ToArray();
```

File: src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs [hunk 3]
```diff
@@ -241,17 +444,28 @@ public void RegisterJavaUDAF(string name, string className)
...
         }
 
         /// <summary>
         /// Helper function to register wrapped udf.
         /// </summary>
-        /// <typeparam name="TResult">Return type of the udf</typeparam>
         /// <param name="name">Name of the udf</param>
         /// <param name="func">Wrapped UDF function</param>
         /// <param name="evalType">The EvalType of the function.</param>
-        internal void Register<TResult>(string name, Delegate func, UdfUtils.PythonEvalType evalType)
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs]: Can we leave this API and just call the new one, so that we don't have to change existing calls?
```C#
internal void Register<TResult>(string name, Delegate func, UdfUtils.PythonEvalType evalType)
{
    return Register(name, func, evalType, UdfUtils.GetReturnType(typeof(TResult));
}
```

File: src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs (hunk 1 above)
Comment (Velociraptor115) [reply src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs]: I have made these changes in the latest commits

File: src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs (hunk 2 above)
Comment (Velociraptor115) [reply src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs]: I have made these changes in the latest commits (used 1st style)

File: src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs (hunk 3 above)
Comment (Velociraptor115) [reply src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs]: I left the previous internal method as it is, but changed the body to delegate to a newer method which accepts the returnType param.
I have also reverted the changes in the two extension files which called the earlier API
Can you please check and verify if this is ok?

File: src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs [hunk 4]
```diff
@@ -241,7 +444,18 @@ public void RegisterJavaUDAF(string name, string className)
         /// <param name="func">Wrapped UDF function</param>
         private void Register<TResult>(string name, Delegate func)
         {
-            Register<TResult>(name, func, UdfUtils.PythonEvalType.SQL_BATCHED_UDF);
+            Register(name, func, UdfUtils.PythonEvalType.SQL_BATCHED_UDF, UdfUtils.GetReturnType(typeof(TResult)));
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs]: Can you revert this change?

File: src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs [hunk 5]
```diff
@@ -252,6 +466,18 @@ private void Register<TResult>(string name, Delegate func)
...
+            Register(name, func, evalType, UdfUtils.GetReturnType(typeof(TResult)));
+        }
+
+        /// <summary>
+        /// Helper function to register wrapped udf.
+        /// </summary>
+        /// <param name="name">Name of the udf</param>
+        /// <param name="func">Wrapped UDF function</param>
+        /// <param name="evalType">The EvalType of the function.</param>
+        /// <param name="returnType">The return type of the function in JSON format.</param>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs]: Can we remove `.` for `<param>`?

File: src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs (hunk 4 above)
Comment (Velociraptor115) [reply src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs]: Sure

File: src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs (hunk 5 above)
Comment (Velociraptor115) [reply src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs]: Sure

File: src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs [hunk 6]
```diff
@@ -244,14 +447,37 @@ private void Register<TResult>(string name, Delegate func)
...
+        }
+
+        /// <summary>
+        /// Helper function to register wrapped udf.
+        /// </summary>
+        /// <param name="name">Name of the udf</param>
+        /// <param name="func">Wrapped UDF function</param>
+        /// <param name="evalType">The EvalType of the function</param>
+        /// <param name="returnType">The return type of the function in JSON format</param>
+        internal void Register(string name, Delegate func, UdfUtils.PythonEvalType evalType, string returnType)
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs]: Can this be `private` for now?

Comment (Velociraptor115) [reply src/csharp/Microsoft.Spark/Sql/UDFRegistration.cs]: Done

**Comment by Velociraptor115** [comment]
Thanks for guiding me through my first PR @imback82 😁

**Comment by imback82** [comment]
Thanks for your first contribution!
